pip/utils/__init__.py

```diff
@@ -104,9 +104,34 @@ def rmtree(dir, ignore_errors=False):
 def rmtree_errorhandler(func, path, exc_info):
     """On Windows, the files in .svn are read-only, so when rmtree() tries to
     remove them, an exception is thrown.  We catch that here, remove the
-    read-only attribute, and hopefully continue without problems."""
+    read-only attribute, and hopefully continue without problems.
+
+    Additionally, in Docker containers with the overlayfs backend, when a
+    package is installed at one layer, and then upgraded or downgraded at
+    another, such that pip uninstalls the previously installed version, an
+    error can occur when files are copied to /tmp (so that they can be rolled
+    back) and then deleted on successful uninstall. See discussion about this
+    issue at https://github.com/docker/docker/issues/9572 and
+    https://github.com/docker/docker/issues/12327. While we are waiting for
+    the overlayfs folks to fix this issue, we need to handle the OSError that
+    happens when we try to delete a file that "isn't there" according to the
+    overlay backend."""
+    if not os.path.exists(path):
+        # no issues with trying to delete a file/directory that isn't there
+        logger.warn("%s(%s) failed because %s does not exist. Ignoring.",
```

Comment:
`logger.warning` is preferred: `warn` is deprecated, cf https://docs.python.org/3/library/logging.html#logging.Logger.warning